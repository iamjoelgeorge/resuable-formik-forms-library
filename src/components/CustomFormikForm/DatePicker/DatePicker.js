import React from 'react';
import PropTypes from 'prop-types';

import CalendarDatePicker from './CalendarDatePicker/CalendarDatePicker';
import DropdownDatePicker from './DropdownDatePicker/DropdownDatePicker';
import { commonProps, commonPropTypes } from '../../../constants/constants';

const DatePicker = (props) => {
  const { isDropdown } = props;

  return (
    <div>{isDropdown ? <DropdownDatePicker {...props} /> : <CalendarDatePicker {...props} />}</div>
  );
};

CalendarDatePicker.propTypes = {
  ...commonPropTypes,
  label: PropTypes.string,
  isDropdown: PropTypes.bool,
  minDate: PropTypes.instanceOf(Date),
  maxDate: PropTypes.instanceOf(Date),
  maxDaysInThePast: PropTypes.number,
  maxDaysInTheFuture: PropTypes.number,
};

CalendarDatePicker.defaultProps = {
  ...commonProps,
  label: '',
  isDropdown: false,
  minDate: null,
  maxDate: new Date('31 Dec 5000'),
  maxDaysInThePast: null,
  maxDaysInTheFuture: null,
};

export default DatePicker;
