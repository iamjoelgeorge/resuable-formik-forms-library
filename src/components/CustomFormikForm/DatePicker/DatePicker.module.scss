@import '../../../theme/variables.scss';
@import '../../../theme/utils.scss';

:global {
  .react-calendar {
    .react-calendar__navigation__arrow {
      &.react-calendar__navigation__next2-button,
      &.react-calendar__navigation__prev2-button {
        display: none;
      }
    }

    .react-calendar__month-view__days {
      padding-top: 10px;

      .react-calendar__tile {
        &.react-calendar__month-view__days__day {
          padding: 10px;
          margin: 2px;
          border: 1px solid transparent !important;

          &:hover {
            border: 1px solid $brightPurple !important;
          }
        }

        &:not(.react-calendar__month-view__days__day--neighboringMonth) {
          background-color: $calendarDayBackground !important;
          cursor: unset !important;

          * {
            background-color: $calendarDayBackground !important;
          }
        }
      }
    }

    .react-calendar__navigation__label {
      padding: 10px;
      font: {
        size: em(13);
        weight: $fontWeightBold;
      }

      .react-calendar__navigation__label__labelText {
        &.react-calendar__navigation__label__labelText--from {
          cursor: pointer;

          &:hover {
            text-decoration: underline;
            text-decoration-color: $brightPurple;
          }
        }
      }
    }
    .react-calendar__year-view__months {
      display: grid !important;
      display: grid;
      grid-template-columns: 1fr 1fr 1fr;
      grid-gap: 2px;

      button.react-calendar__tile {
        &.react-calendar__year-view__months__month {
          min-width: 100%;
          font-size: em(13);
          height: 44px;
          font-weight: $fontWeightBold;
          cursor: pointer;

          &:hover {
            * {
              background-color: $lightPurple !important;
            }
            color: $darkPurple;
            background-color: $lightPurple !important;
          }
        }
      }
    }

    .react-calendar__century-view__decades,
    .react-calendar__decade-view__years {
      button {
        font-size: em(13);
        min-width: 100%;
        height: 44px;
        font-weight: $fontWeightBold;
        cursor: pointer;

        &:hover {
          color: $darkPurple;
          background-color: $lightPurple !important;

          * {
            background-color: $lightPurple !important;
          }
        }
      }
    }

    .react-calendar__viewContainer {
      max-height: 271px;
      overflow: scroll;
    }
  }
}

.container {
  max-width: 279px;
  position: relative;
  cursor: pointer;
  margin-top: 30px;
  z-index: 1;

  .selectedDateContainer {
    border-bottom: 1px solid $lighterGrey;

    &:hover {
      border-bottom-color: $brightPurpleTwo;
    }

    .selectedDate {
      color: $secondaryGrey;
      font-size: em(14);
      margin: 2px 0 5px 10px;
    }
  }

  .calendarContainer {
    position: absolute;
    top: 100%;
    left: 0;
    width: 100%;
    padding: 10px;
    background: $white;
    box-shadow: 0 2px 2px $borderGrey;

    &:hover {
      & + .selectedDateContainer {
        border-bottom-color: $primary;
      }
    }

    * {
      background-color: $white;
      border: none;
    }
  }
}
