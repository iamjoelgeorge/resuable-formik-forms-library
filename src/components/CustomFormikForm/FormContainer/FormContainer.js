import React, { Children, cloneElement } from 'react';

import PropTypes from 'prop-types';
import { Formik, Form } from 'formik';
import * as Yup from 'yup';

import './FormContainer.module.scss';
import { validateInput } from '../../../utils/validateInput';

const FormContainer = (props) => {
  const { initialValues, children, validations, containerClass, onSubmit, ...rest } = props;

  const validationObject = {};
  validations.forEach((item) => {
    validateInput(item, validationObject);
  });

  const validationSchema = Yup.object().shape(validationObject);

  const renderChildren = (child, formik) => {
    const childOfChild = child.props.children;
    const { className } = child.props;

    if (childOfChild?.props?.children) {
      return renderChildren(childOfChild, formik);
    }

    return (
      <div className={className}>
        {childOfChild?.length
          ? /*
            The error generated because of the key not being added can be ignored in this case
            because these elements will not be deleted/modified. The key is not added the the child
            because it is interfering with the focus event of the input field.
          */
            childOfChild?.map((childElement) => cloneElement(childElement, { formik }))
          : cloneElement(childOfChild, { formik })}
      </div>
    );
  };
  return (
    <Formik initialValues={initialValues} onSubmit={onSubmit} validationSchema={validationSchema}>
      {(formik) => {
        const childrenWithExtraProp = Children.map(children, (child) => {
          return child?.props?.children
            ? renderChildren(child, formik)
            : cloneElement(child, { formik });
        });
        return (
          <Form className={containerClass} {...rest}>
            {childrenWithExtraProp}
          </Form>
        );
      }}
    </Formik>
  );
};

FormContainer.propTypes = {
  initialValues: PropTypes.shape({}),
  children: PropTypes.element,
  validations: PropTypes.array,
  containerClass: PropTypes.string,
  onSubmit: PropTypes.func,
};

FormContainer.defaultProps = {
  initialValues: {},
  children: null,
  validations: [],
  containerClass: '',
  onSubmit: () => {},
};

export default FormContainer;
