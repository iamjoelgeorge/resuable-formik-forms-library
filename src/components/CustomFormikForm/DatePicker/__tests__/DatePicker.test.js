/* eslint-disable */
import React from 'react';

import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import renderer from 'react-test-renderer';

import FormContainer from '../../FormContainer/FormContainer';
import DatePicker from '../DatePicker';

const defaultProps = {
  name: 'departureDate',
  label: 'Departure Date',
  isDropdown: false,
};

const initialValues = {
  departureDate: null,
};

const defaultValidations = [];
const handleSubmit = jest.fn();

afterEach(cleanup);

describe('[Component]: DatePicker', () => {
  it('should match the Snapshot', () => {
    const tree = renderer
      .create(
        <FormContainer
          initialValues={initialValues}
          validations={defaultValidations}
          onSubmit={handleSubmit}
        >
          <DatePicker {...defaultProps} />
        </FormContainer>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render the Calendar Date Picker', () => {
    const { getByTestId } = render(
      <FormContainer
        initialValues={initialValues}
        validations={defaultValidations}
        onSubmit={handleSubmit}
      >
        <DatePicker {...defaultProps} />
      </FormContainer>,
    );

    expect(getByTestId('departureDate-calendar-datepicker')).toBeTruthy();
  });

  it('should render the Dropdown Date Picker', () => {
    const props = {
      ...defaultProps,
      isDropdown: true,
    };
    const { getByTestId } = render(
      <FormContainer
        initialValues={initialValues}
        validations={defaultValidations}
        onSubmit={handleSubmit}
      >
        <DatePicker {...props} />
      </FormContainer>,
    );

    expect(getByTestId('departureDate-dropdown-datepicker')).toBeTruthy();
  });
});
