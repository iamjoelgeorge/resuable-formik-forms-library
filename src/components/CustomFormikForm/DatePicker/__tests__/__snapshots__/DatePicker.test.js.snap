// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Component]: DatePicker should match the Snapshot 1`] = `
<form
  action="#"
  className=""
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div>
    <div
      className="container "
      data-testid="departureDate-calendar-datepicker"
      id="custom-calendar-date-picker"
    >
      <div>
        <button
          className="selectedDateContainer"
          data-testid="departureDate-calendar-toggle-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <div
            className="container  "
          >
            <label
              className="label"
              htmlFor="departureDate-selectedDate"
            >
              Departure Date
            </label>
          </div>
          <p
            className="selectedDate"
            data-testid="departureDate-selectedDate"
            id="departureDate-selectedDate"
          >
            <span
              className="dropdownIcon"
            >
              <img
                alt="Dropdown icon"
                src="test-file-stub"
              />
            </span>
          </p>
        </button>
      </div>
      
      
      
    </div>
  </div>
</form>
`;
